File: lua/vimficiency/sequence_display.lua
local config = require("vimficiency.config")
local ffi_display = require("vimficiency.ffi.display")

local M = {}

local function resolved_opts(opts)
  local defaults = config.sequence_display or {}
  opts = opts or {}
  local tokenize = defaults.tokenize ~= false
  local sectionize = defaults.sectionize ~= false
  if opts.tokenize ~= nil then
    tokenize = opts.tokenize
  end
  if opts.sectionize ~= nil then
    sectionize = opts.sectionize
  end
  return {
    tokenize = tokenize,
    sectionize = sectionize,
  }
end

local function display_token_text(token)
  return token.kind == "typed"
    and ffi_display.sequence_display_typed_text(token.text, true)
    or ffi_display.sequence_display_key_notation(token.text)
end

local function join_token_chunks(tokens, tokenize)
  local chunks = {}
  for i, token in ipairs(tokens) do
    local prev = tokens[i - 1]
    if tokenize and i > 1 and not (prev.kind == "typed" and token.kind == "typed") then
      chunks[#chunks + 1] = { " ", "Normal" }
    end
    chunks[#chunks + 1] = { token.text, token.hl or "Normal" }
  end
  return chunks
end

local section_tokens

local function chunked_token_lines(tokens, tokenize, sectionize)
  if sectionize then
    local sections = section_tokens(tokens)
    if #sections > 1 then
      local lines = {}
      for _, section in ipairs(sections) do
        lines[#lines + 1] = join_token_chunks(section.tokens, tokenize)
      end
      return lines
    end
  end

  return { join_token_chunks(tokens, tokenize) }
end

function section_tokens(tokens)
  local sections = {}
  local current_movement = {}
  local edit_start = { change = true, delete = true }
  local edit_continuation = {
    change = true,
    delete = true,
    typed = true,
    escape = true,
  }

  local function flush_movement()
    if #current_movement == 0 then return end
    sections[#sections + 1] = {
      kind = "movement",
      tokens = current_movement,
    }
    current_movement = {}
  end

  local i = 1
  while i <= #tokens do
    local token = tokens[i]
    if edit_start[token.kind] then
      flush_movement()
      local edit_tokens = {}
      while i <= #tokens and edit_continuation[tokens[i].kind] do
        edit_tokens[#edit_tokens + 1] = tokens[i]
        i = i + 1
      end
      sections[#sections + 1] = {
        kind = "edit",
        tokens = edit_tokens,
      }
    else
      current_movement[#current_movement + 1] = token
      i = i + 1
    end
  end

  flush_movement()
  return sections
end

function M.lines(seq, opts)
  return ffi_display.sequence_display_lines(seq, opts)
end

function M.inline(seq, opts)
  return ffi_display.sequence_display_inline(seq, opts)
end

---@param tokens VF.Sequence.Token[]
---@param opts { tokenize?: boolean, sectionize?: boolean, highlight_index?: integer, highlight_group?: string }?
---@return table[][]
function M.chunked_lines_for_tokens(tokens, opts)
  local resolved = resolved_opts(opts)
  opts = opts or {}

  if not tokens or #tokens == 0 then
    return { { { "", "Normal" } } }
  end

  local display_tokens = {}
  for i, token in ipairs(tokens) do
    local index = token.index or i
    display_tokens[#display_tokens + 1] = {
      text = display_token_text(token),
      kind = token.kind,
      index = index,
      hl = opts.highlight_index == index
        and (opts.highlight_group or "Visual")
        or "Normal",
    }
  end

  return chunked_token_lines(display_tokens, resolved.tokenize, resolved.sectionize)
end

---@param text string
---@param opts { tokenize?: boolean }?
---@return string
function M.typed_text_inline(text, opts)
  return ffi_display.sequence_display_typed_text(text, true)
end

---@param text string
---@param opts { tokenize?: boolean }?
---@return string
function M.literal_typed_text_inline(text, opts)
  return ffi_display.sequence_display_typed_text(text, false)
end

---@param chunks { kind: string, text: string }[]
---@param opts { tokenize?: boolean }?
---@return string
function M.typed_chunks_inline(chunks, opts)
  return ffi_display.sequence_display_typed_chunks(chunks, opts)
end

function M.prefixed_lines(prefix, seq, opts, suffix)
  local rendered = M.lines(seq, opts)
  local first = rendered[1] or ""
  local out = { prefix .. first .. (suffix or "") }
  local continuation = string.rep(" ", #prefix)
  for i = 2, #rendered do
    out[#out + 1] = continuation .. rendered[i]
  end
  return out
end

return M

File: src/Interpreter/SequenceDisplay.cpp
#include "Interpreter/SequenceDisplay.h"

#include "Interpreter/SequenceParser.h"
#include "Utils/PrettyText.h"

#include <string>

using namespace std;

namespace VF::SequenceDisplay {
namespace {

string glyphForKeyNotation(string_view special) {
  if (special == "<Space>") return string(PrettyText::SPACE);
  if (special == "<Tab>") return string(PrettyText::TAB);
  if (special == "<CR>" || special == "<Enter>" || special == "<Return>") {
    return string(PrettyText::NEWLINE);
  }
  if (special == "<lt>" || special == "<LT>") return "<";
  return string(special);
}

vector<string_view> splitKeyNotationTypedText(string_view text) {
  vector<string_view> parts;
  size_t plainStart = 0;
  size_t i = 0;
  while (i < text.size()) {
    if (text[i] != '<') {
      i++;
      continue;
    }

    const size_t close = text.find('>', i);
    if (close == string_view::npos) {
      i++;
      continue;
    }

    if (plainStart < i) {
      parts.push_back(text.substr(plainStart, i - plainStart));
    }
    parts.push_back(text.substr(i, close - i + 1));
    i = close + 1;
    plainStart = i;
  }

  if (plainStart < text.size()) {
    parts.push_back(text.substr(plainStart));
  }
  return parts;
}

string displayTokenText(const TaggedToken& token) {
  return token.kind == TokenKind::TypedText
      ? displayTypedText(token.token)
      : displayKeyNotation(token.token);
}

struct DisplayToken {
  string text;
  TokenKind kind;
};

bool isEditStart(TokenKind kind) {
  return kind == TokenKind::Change || kind == TokenKind::Delete;
}

bool isEditContinuation(TokenKind kind) {
  return kind == TokenKind::Change ||
         kind == TokenKind::Delete ||
         kind == TokenKind::TypedText ||
         kind == TokenKind::Escape;
}

string joinTokens(const vector<DisplayToken>& tokens, bool tokenize) {
  string out;
  for (size_t i = 0; i < tokens.size(); i++) {
    if (tokenize && i > 0) out += ' ';
    out += tokens[i].text;
  }
  return out;
}

vector<vector<DisplayToken>> sectionTokens(const vector<DisplayToken>& tokens) {
  vector<vector<DisplayToken>> sections;
  vector<DisplayToken> currentMovement;

  auto flushMovement = [&]() {
    if (currentMovement.empty()) return;
    sections.push_back(std::move(currentMovement));
    currentMovement = {};
  };

  size_t i = 0;
  while (i < tokens.size()) {
    if (!isEditStart(tokens[i].kind)) {
      currentMovement.push_back(tokens[i]);
      i++;
      continue;
    }

    flushMovement();
    vector<DisplayToken> editTokens;
    while (i < tokens.size() && isEditContinuation(tokens[i].kind)) {
      editTokens.push_back(tokens[i]);
      i++;
    }
    sections.push_back(std::move(editTokens));
  }

  flushMovement();
  return sections;
}

vector<string> tokenizedLines(string_view seq, Options opts) {
  if (seq.empty()) {
    return {""};
  }

  auto parsed = parseSequence(seq);
  if (!parsed || parsed->empty()) {
    return {displayKeyNotation(seq)};
  }

  vector<DisplayToken> tokens;
  tokens.reserve(parsed->size());
  for (const TaggedToken& token : *parsed) {
    tokens.push_back(DisplayToken{
        .text = displayTokenText(token),
        .kind = token.kind,
    });
  }

  if (opts.sectionize) {
    const vector<vector<DisplayToken>> sections = sectionTokens(tokens);
    if (sections.size() > 1) {
      vector<string> out;
      out.reserve(sections.size());
      for (const auto& section : sections) {
        out.push_back(joinTokens(section, opts.tokenize));
      }
      return out;
    }
  }

  return {joinTokens(tokens, opts.tokenize)};
}

}  // namespace

string displayKeyNotation(string_view text) {
  string out;
  size_t i = 0;
  while (i < text.size()) {
    if (text[i] != '<') {
      out += text[i];
      i++;
      continue;
    }

    const size_t close = text.find('>', i);
    if (close == string_view::npos) {
      out += text[i];
      i++;
      continue;
    }

    out += glyphForKeyNotation(text.substr(i, close - i + 1));
    i = close + 1;
  }
  return out;
}

string displayTypedText(string_view text) {
  string out;
  for (const string_view part : splitKeyNotationTypedText(text)) {
    const bool keyNotation =
        part.size() >= 2 && part.front() == '<' && part.back() == '>';
    out += keyNotation ? displayKeyNotation(part) : prettify(part);
  }
  return out;
}

string displayLiteralTypedText(string_view text) {
  return prettify(text);
}

vector<string> lines(string_view seq, Options opts) {
  if (!opts.tokenize && !opts.sectionize) {
    return {string(seq)};
  }
  return tokenizedLines(seq, opts);
}

string inlineText(string_view seq, Options opts) {
  opts.sectionize = false;
  const vector<string> rendered = tokenizedLines(seq, opts);
  return rendered.empty() ? "" : rendered.front();
}

string typedChunksInline(const vector<TypedChunk>& chunks, bool tokenize) {
  vector<DisplayToken> tokens;
  tokens.reserve(chunks.size());
  for (const TypedChunk& chunk : chunks) {
    string text = chunk.kind == "literal"
        ? displayLiteralTypedText(chunk.text)
        : displayKeyNotation(chunk.text);
    if (!text.empty()) {
      tokens.push_back(DisplayToken{
          .text = std::move(text),
          .kind = TokenKind::TypedText,
      });
    }
  }

  return joinTokens(tokens, tokenize);
}

vector<string> prefixedLines(
    string_view prefix,
    string_view seq,
    Options opts,
    string_view suffix) {
  vector<string> rendered = lines(seq, opts);
  const string first = rendered.empty() ? "" : rendered.front();
  vector<string> out;
  out.push_back(string(prefix) + first + string(suffix));

  const string continuation(prefix.size(), ' ');
  for (size_t i = 1; i < rendered.size(); i++) {
    out.push_back(continuation + rendered[i]);
  }
  return out;
}

}  // namespace VF::SequenceDisplay
